File: Final_Project/Code/Project_4/Bai_toan_5/Exercises/7/CSS_bt5_ex7.py
from collections import deque

class Tree:
    def __init__(self, n):
        # n: so nut
        # adj: danh sach ke
        self.n = n
        self.adj = [[] for _ in range(n)]

    def add_edge(self, u, v):
        # Them canh (u,v)
        self.adj[u].append(v)

    def generate_complete_binary(self):
        # Sinh cay nhi phan day du
        if self.n == 0:
            return
        queue = deque([0])
        i = 1
        while queue and i < self.n:
            u = queue.popleft()
            if i < self.n:
                self.add_edge(u, i)
                queue.append(i)
                i += 1
            if i < self.n:
                self.add_edge(u, i)
                queue.append(i)
                i += 1

# Vi du su dung
if __name__ == "__main__":
    t = Tree(7)
    t.generate_complete_binary()
    print("Complete binary tree edges:", 
          [(u, v) for u in range(t.n) for v in t.adj[u]])

File: Final_Project/Code/Project_4/Bai_toan_5/Exercises/7/CSS_bt5_ex7.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class Tree {
public:
    int n;                   // so nut
    vector<vector<int>> adj; // danh sach ke

    Tree(int vertices) : n(vertices) {
        adj.resize(n);
    }

    void add_edge(int u, int v) {
        // Them canh (u,v)
        adj[u].push_back(v);
    }

    void generate_complete_binary() {
        // Sinh cay nhi phan day du
        if (n == 0) return;
        queue<int> q;
        q.push(0);
        int i = 1;
        while (!q.empty() && i < n) {
            int u = q.front(); q.pop();
            if (i < n) {
                add_edge(u, i);
                q.push(i);
                i++;
            }
            if (i < n) {
                add_edge(u, i);
                q.push(i);
                i++;
            }
        }
    }
};

int main() {
    Tree t(7);
    t.generate_complete_binary();
    cout << "Complete binary tree edges: ";
    for (int u = 0; u < t.n; ++u) {
        for (int v : t.adj[u]) {
            cout << "(" << u << "," << v << ") ";
        }
    }
    cout << endl;
    return 0;
}